File: ts/src/catalog.ts
import { existsSync, readFileSync } from "node:fs";
import { fileURLToPath } from "node:url";
import { dirname, resolve } from "node:path";
import type { ModelSpec, ProviderName } from "./types.js";
import { EFFORT_LADDER, TASKS } from "./types.js";
import { ENV, has } from "./env.js";

const here = dirname(fileURLToPath(import.meta.url));

/** The shared catalog lives at the repo root; fall back to a package-local copy when published. */
function defaultCatalogPath(): string {
  const candidates = [
    resolve(process.cwd(), "models.json"),
    resolve(here, "..", "..", "models.json"),
    resolve(here, "..", "models.json"),
  ];
  return candidates.find((p) => existsSync(p)) ?? candidates[0];
}

/** Load the default catalog shipped with frugal, or a user-supplied JSON file. */
export function loadCatalog(path?: string): ModelSpec[] {
  const file = path ?? process.env[ENV.catalog] ?? defaultCatalogPath();
  const raw = JSON.parse(readFileSync(file, "utf8")) as { models?: ModelSpec[] } | ModelSpec[];
  const models = Array.isArray(raw) ? raw : (raw.models ?? []);
  for (const m of models) validate(m);
  return models;
}

function validate(m: ModelSpec): void {
  const bad = (msg: string): never => {
    throw new Error(`catalog: model "${m?.id ?? "?"}": ${msg}`);
  };
  if (!m.id || !m.provider || !m.model) bad("id, provider and model are required");
  if (![1, 2, 3].includes(m.tier)) bad("tier must be 1, 2 or 3");
  if (!m.price || m.price.input < 0 || m.price.output < 0) bad("price.input/output required");
  for (const e of m.effort ?? []) if (!EFFORT_LADDER.includes(e)) bad(`unknown effort "${e}"`);
  for (const t of [...(m.strengths ?? []), ...(m.weaknesses ?? [])])
    if (!TASKS.includes(t)) bad(`unknown task "${t}"`);
}

const PROVIDER_KEY: Record<ProviderName, string> = {
  anthropic: ENV.anthropic,
  openai: ENV.openai,
  openrouter: ENV.openrouter,
};

/** Models whose provider has credentials configured. */
export function availableModels(catalog: ModelSpec[]): ModelSpec[] {
  return catalog.filter((m) => has(PROVIDER_KEY[m.provider]));
}

export function findModel(catalog: ModelSpec[], name: string): ModelSpec | undefined {
  const n = name.toLowerCase();
  return catalog.find((m) => m.id.toLowerCase() === n || m.model.toLowerCase() === n);
}

File: csharp/Dispatcher/Catalog.cs
using System.Text.Json;

namespace LlmDispatcher;

public static class Catalog
{
    private static readonly Dictionary<string, string> ProviderKey = new()
    {
        ["anthropic"] = Env.Anthropic,
        ["openai"] = Env.OpenAI,
        ["openrouter"] = Env.OpenRouter,
    };

    /// <summary>The shared catalog lives at the repo root; search there, then next to the binary.</summary>
    public static string DefaultPath()
    {
        var candidates = new List<string>();
        var dir = Directory.GetCurrentDirectory();
        for (var i = 0; i < 4 && dir is not null; i++)
        {
            candidates.Add(Path.Combine(dir, "models.json"));
            dir = Path.GetDirectoryName(dir);
        }
        candidates.Add(Path.Combine(AppContext.BaseDirectory, "models.json"));
        return candidates.FirstOrDefault(File.Exists) ?? candidates[0];
    }

    public static List<ModelSpec> Load(string? path = null)
    {
        var file = path ?? Env.Get(Env.Catalog) ?? DefaultPath();
        using var doc = JsonDocument.Parse(File.ReadAllText(file), new JsonDocumentOptions { CommentHandling = JsonCommentHandling.Skip });
        var root = doc.RootElement;
        var arr = root.ValueKind == JsonValueKind.Array ? root : root.GetProperty("models");
        var models = arr.Deserialize<List<ModelSpec>>(Json.Camel) ?? [];
        foreach (var m in models) Validate(m);
        return models;
    }

    private static void Validate(ModelSpec m)
    {
        void Bad(string msg) => throw new InvalidOperationException($"catalog: model \"{m.Id}\": {msg}");
        if (string.IsNullOrEmpty(m.Id) || string.IsNullOrEmpty(m.Provider) || string.IsNullOrEmpty(m.Model)) Bad("id, provider and model are required");
        if (m.Tier is < 1 or > 3) Bad("tier must be 1, 2 or 3");
        if (m.Price.Input < 0 || m.Price.Output < 0) Bad("price.input/output required");
        if (!ProviderKey.ContainsKey(m.Provider)) Bad($"unknown provider \"{m.Provider}\"");
        foreach (var e in m.Effort) if (!Efforts.IsValid(e)) Bad($"unknown effort \"{e}\"");
        foreach (var t in (m.Strengths ?? []).Concat(m.Weaknesses ?? [])) if (!Tasks.All.Contains(t)) Bad($"unknown task \"{t}\"");
    }

    /// <summary>Models whose provider has credentials configured.</summary>
    public static List<ModelSpec> Available(IEnumerable<ModelSpec> catalog) =>
        catalog.Where(m => Env.Has(ProviderKey[m.Provider])).ToList();

    public static ModelSpec? Find(IEnumerable<ModelSpec> catalog, string name) =>
        catalog.FirstOrDefault(m =>
            string.Equals(m.Id, name, StringComparison.OrdinalIgnoreCase) ||
            string.Equals(m.Model, name, StringComparison.OrdinalIgnoreCase));
}
